optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -180,7 +190,332 @@
      * @return a stream that matches every combination of A and B for which all the {@link BiJoiner joiners} are true
      */
     default <B> BiConstraintStream<A, B> join(Class<B> otherClass, BiJoiner<A, B>... joiners) {
-        return join(otherClass, AbstractBiJoiner.merge(joiners));
+        int joinerCount = joiners.length;
+        int indexOfFirstFilter = -1;
+        for (int i = 0; i < joinerCount; i++) {
+            BiJoiner<A, B> joiner = joiners[i];
+            boolean hasAFilter = indexOfFirstFilter >= 0;
+            if (joiner instanceof FilteringBiJoiner) {
+                if (!hasAFilter) { // From now on, we only allow filtering joiners.
+                    indexOfFirstFilter = i;
+                }
+            } else if (hasAFilter) {
+                throw new IllegalStateException("Indexing joiner (" + joiner + ") must not follow a filtering joiner ("
+                        + joiners[indexOfFirstFilter] + ").");
+            }
+        }
+        if (indexOfFirstFilter < 0) { // Only found indexing joiners.
+            return join(otherClass, AbstractBiJoiner.merge(joiners));
+        }
+        // Assemble the join stream.
+        BiConstraintStream<A, B> joined = indexOfFirstFilter == 0 ?
+                join(otherClass) :
+                join(otherClass, Arrays.copyOf(joiners, indexOfFirstFilter));
+        // We merge all filters into one, so that we don't pay the penalty for lack of indexing more than once.
+        BiPredicate<A, B> resultingFilter = (a, b) -> true;
```

Comment:
Don't create this "always true" predicate:

- If there is no filtering, don't call filter().
- If there is only 1 filter, let's say `Foo::isReady`, don't create a wrapper predicate of an always true predicate combined with foo.isReady(), just use `Foo::isReady``directly.
- Same reasoning for 2 filters (don't make it 3)
- ...

Reply:
Motivation:
- performance (especially with the extra filter() call for those that don't use it, it currently will cause non-trivial perf degredation in Bavet that creates tuples for filters)
- node sharing (same lambda recognition)